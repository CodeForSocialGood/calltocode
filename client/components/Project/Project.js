import React, { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { GridListTile, GridListTileBar } from 'material-ui/GridList'

import styles from './Project.scss'

class Project extends Component {
  constructor (props) {
    super(props)

    this.handleClick = this.handleClick.bind(this)
    this.renderProjectApplicationResult = this.renderProjectApplicationResult.bind(this)
    this.getAppliedStatus = this.getAppliedStatus.bind(this)
  }

  handleClick () {
    const { authenticated, project, user } = this.props
    const applied = this.getAppliedStatus()

    if (!applied && authenticated && user.usertype === 'volunteer') {
      return this.props.applyForProject(project, user)
    }
  }

  renderProjectApplicationResult (project) {
    const applied = this.getAppliedStatus()
    const isContact = this.isUserTypeContact()
    if (applied && !isContact) {
      return (
        <span className={styles.projectApplyPass}>&#10004;</span>
      )
    } else if (!isContact) {
      return (
        <span className={styles.projectApplyFail}>&#10007;</span>
      )
    }
  }

  getAppliedStatus () {
    return (
      this.props.user.projectsAppliedFor &&
      this.props.user.projectsAppliedFor.includes(this.props.project.id)
    )
  }

  isUserTypeContact () {
    return this.props.user.usertype === 'contact'
  }

  render () {
    const { project } = this.props

    const applied = this.getAppliedStatus()
    const isContact = this.isUserTypeContact()
    let projectClasses = styles.project
    if (this.props.authenticated && !applied && !isContact) {
      projectClasses = styles.projectAuthenticated
    } else if (this.props.authenticated && !applied) {
      projectClasses = styles.projectAuthenticatedContact
    }

    return (
      <GridListTile className={projectClasses}
        onClick={this.handleClick.bind(this)}>
        <img src={project.image || require('../../images/logo.png')} />
        <GridListTileBar title={project.name}
          subtitle={project.organization.name || 'Organization Name'}
          actionIcon={this.renderProjectApplicationResult(project)}>
        </GridListTileBar>
      </GridListTile>
    )
  }
}

function mapStateToProps (state) {
  return {
    user: state.user
  }
}

Project.propTypes = {
  applyForProject: PropTypes.func.isRequired,
  authenticated: PropTypes.bool.isRequired,
  project: PropTypes.object.isRequired,
  user: PropTypes.object
}

export default connect(mapStateToProps, null)(Project)
