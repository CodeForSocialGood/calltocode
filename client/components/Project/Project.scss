@import '../../scss/partials/partials';

.project {
  border-radius: 5px;
  color: white;
}

.projectAuthenticated,
.projectAuthenticatedContact {
  @extend .project;
}

.projectAuthenticated:hover {
  @extend %hover;
  @extend %boxShadowSmall;
}

.projectClasses {
  display: grid;
  grid-template-areas: 
    "img content";
}

.projectContent {
  grid-area: content;
}

.image {
  grid-area: img;
  width: 100%;
}

.projectApply {
  margin-right: 15px;
}

.projectApplyPass {
  @extend .projectApply;
  color: green;
}

.projectApplyFail {
  @extend .projectApply;
  color: red;
}

