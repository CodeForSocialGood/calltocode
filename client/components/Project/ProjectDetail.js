import React, { Component } from 'react'
import { connect } from 'react-redux'
import ProjectActionCreator from '../../actions/project'
import PropTypes from 'prop-types'
import {Link} from 'react-router-dom'
import Card, {CardHeader, CardContent, CardMedia} from 'material-ui/Card'
import Typography from 'material-ui/Typography';


class ProjectDetail extends Component {
  constructor (args) {
    super(args)
    this.state = { projectId: this.props.match.params.id }
  }

  componentDidMount () {
    this.props.fetchProjectById(this.state.projectId)
  }

  renderApplication (application) {
    return (
      <Card>
        <CardHeader
          title= {application.volunteer.email}
        />
        <CardContent>
          <Typography>
           message: {application.message}
          </Typography>

          <Typography>
            Submitted at: {application.createdAt}
          </Typography>
        </CardContent>
      </Card>
    )
  }

  render () {
    const project = this.props.project || {}
    const applications = project.applications || []
    if (applications.length === 0) {
      return (
        <div>
          <h1>No one applied this project</h1>
          <Link to={`/profile`}>Go Back</Link>
        </div>)
    } else {
      return (
        <div>
          <h1>The following applicants applied this project:</h1>
          <ul>
            {
              applications.map((app) => <li key={app.id}> {this.renderApplication(app)} </li>)
            }
          </ul>
          <Link to={`/profile`} >Go Back</Link>
        </div>
      )
    }
  }
}

const mapDispatchToProps = {
  fetchProjectById: ProjectActionCreator.fetchProjectById
}

const mapStateToProperties = (state) => {
  return {project: state.project.project || {}}
}

ProjectDetail.propTypes = {
  project: PropTypes.object,
  fetchProjectById: PropTypes.func,
  match: PropTypes.object
}

export default connect(mapStateToProperties, mapDispatchToProps)(ProjectDetail)
