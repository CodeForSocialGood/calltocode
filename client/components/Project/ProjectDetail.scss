@import '../../scss/partials/partials';

.goBack {
  margin-top: 20px;
}
