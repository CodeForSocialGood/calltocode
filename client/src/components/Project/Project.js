import React, { Component } from 'react'
import PropTypes from 'prop-types'
import styles from '../ListOfProjects/ListOfProjects.scss'
import emailApiClient from '../../api/email'

class Project extends Component {
  constructor (props) {
    super(props)

    this.handleClick = this.handleClick.bind(this)
    this.mailToOrganization = this.mailToOrganization.bind(this)
    this.renderProjectApplicationResult = this.renderProjectApplicationResult.bind(this)
  }

  handleClick () {
    if (this.props.loggedIn /* TODO && ! this.props.project.applied */) {
      return this.mailToOrganization(this.props.project)
    }
  }

  mailToOrganization (project) {
    return emailApiClient.send(project).then(response => {
      console.log(project)
      this.props.applyForProject(project)
    })
  }

  renderProjectApplicationResult (project) {
    if (this.props.project.applicationResult === true) {
      return (
        <span className={styles.listApplyPass}>
          &#10004;
        </span>
      )
    } else if (this.props.project.applicationResult === false) {
      return (
        <span className={styles.listApplyFail}>
          &#10007;
        </span>
      )
    } else { return null }
  }

  render () {
    // const applied = this.props.project.applicationResult === true || this.props.project.applicationResult === false
    const liClassName = this.props.loggedIn /* TODO && !applied */ ? styles.listOrgLoggedIn : styles.listOrg

    return (
      <li
        onClick={this.handleClick.bind(this)}>
        {this.renderProjectApplicationResult(this.props.project)}
        <div className={liClassName}>
          Name:{this.props.project.name} Role:{this.props.project.role}
        </div>
      </li>
    )
  }
}

Project.propTypes = {
  loggedIn: PropTypes.bool.isRequired,
  project: PropTypes.object.isRequired,
  applyForProject: PropTypes.func.isRequired
}

export default Project
