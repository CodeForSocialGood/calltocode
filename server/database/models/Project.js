import mongoose from 'mongoose'

const ProjectSchema = mongoose.Schema({
  organization: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Organization'
  },
  name: String,
  role: String,
  email: String
}, { timestamps: true })

ProjectSchema.methods.toJSON = function () {
  return {
    id: this._id,
    organization: this.organization,
    name: this.name,
    role: this.role,
    email: this.email,
    createdAt: this.createdAt,
    updatedAt: this.updatedAt
  }
}

export default mongoose.model('Project', ProjectSchema)
