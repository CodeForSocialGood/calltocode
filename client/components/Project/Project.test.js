/* eslint-env jest */
import React from 'react'
import {shallow} from 'enzyme'

import {Project} from './Project'
let mockProps

beforeEach(() => {
  mockProps = {
    applyForProject: () => {},
    authenticated: null,
    project: {
      name: '',
      organization: {name: ''},
      image: '',
      id: 1
    },
    user: {
      usertype: null,
      projectsAppliedFor: [{type: 1, ref: ''}],
      email: '',
      salt: '',
      hash: '',
      organization: {type: 1, ref: ''}
    }
  }
})

describe('<Project />', () => {
  describe('authenticated is truthy', () => {
    beforeEach(() => { mockProps.autheticated = true })

    describe('user has applied for project', () => {
      beforeEach(() => {
        mockProps.user = {
          usertype: null,
          projectsAppliedFor: [1]
        }
      })

      describe('user\'s usertype is "contact"', () => {
        beforeEach(() => { mockProps.user.usertype = 'contact' })

        test('component renders correctly', () => {
          const component = shallow(<Project {...mockProps} />)
          expect(component).toMatchSnapshot()
        })
      })

      describe('user\'s usertype is "volunteer"', () => {
        beforeEach(() => { mockProps.user.usertype = 'volunteer' })

        const component = shallow(<Project {...mockProps} />)
        expect(component).toMatchSnapshot()
      })
    })

    describe('user has not applied for project', () => {
      beforeEach(() => {
        mockProps.user = {
          usertype: null,
          projectsAppliedFor: []
        }
      })

      describe('user\'s usertype is "volunteer"', () => {
        beforeEach(() => { mockProps.user.usertype = 'volunteer' })
        // component for applyForProject renders
        const component = shallow(<Project {...mockProps} />)
        expect(component).toMatchSnapshot()
      })
      describe('user\'s usertype is "contact"', () => {
        beforeEach(() => { mockProps.user.usertype = 'contact' })

        const component = shallow(<Project {...mockProps} />)
        expect(component).toMatchSnapshot()
      })
    })
  })

  // describe('authenticated is falsey', () => {
  //   beforeEach(() => { mockProps.autheticated = true })
  //
  //   const component = shallow(<Project {...mockProps} />)
  //   expect(component).toMatchSnapshot()
  // })
})
