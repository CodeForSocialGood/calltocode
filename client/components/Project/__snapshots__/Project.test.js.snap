// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Project /> authenticated is truthy user has applied for project user's usertype is "volunteer" component renders correclty 1`] = `
<WithStyles(GridListTile)
  className="project"
  onClick={[Function]}
>
  <img
    src={Object {}}
  />
  <WithStyles(GridListTileBar)
    actionIcon={
      <span
        className="projectApplyPass"
      >
        ✔
      </span>
    }
    subtitle="Organization Name"
    title=""
  />
</WithStyles(GridListTile)>
`;

exports[`<Project /> authenticated is truthy user has not applied for project user's usertype is "contact" component renders correclty 1`] = `
<WithStyles(GridListTile)
  className="project"
  onClick={[Function]}
>
  <img
    src={Object {}}
  />
  <WithStyles(GridListTileBar)
    subtitle="Organization Name"
    title=""
  />
</WithStyles(GridListTile)>
`;

exports[`<Project /> authenticated is truthy user has not applied for project user's usertype is "volunteer" component renders correclty 1`] = `
<WithStyles(GridListTile)
  className="project"
  onClick={[Function]}
>
  <img
    src={Object {}}
  />
  <WithStyles(GridListTileBar)
    actionIcon={
      <span
        className="projectApplyFail"
      >
        ✗
      </span>
    }
    subtitle="Organization Name"
    title=""
  />
</WithStyles(GridListTile)>
`;
